modules/renter/hostdb/hostdb.go

```diff
@@ -52,31 +51,27 @@ type HostDB struct {
 
 	// The hostTree is the root node of the tree that organizes hosts by
 	// weight. The tree is necessary for selecting weighted hosts at
-	// random. 'activeHosts' provides a lookup from hostname to the the
-	// corresponding node, as the hostTree is unsorted. A host is active if
-	// it is currently responding to queries about price and other
-	// settings.
-	hostTree    *hosttree.HostTree
-	activeHosts map[modules.NetAddress]*hostEntry
-
-	// allHosts is a simple list of all known hosts by their network address,
-	// including hosts that are currently offline.
-	allHosts map[modules.NetAddress]*hostEntry
+	// random.
+	hostTree *hosttree.HostTree
 
 	// the scanPool is a set of hosts that need to be scanned. There are a
 	// handful of goroutines constantly waiting on the channel for hosts to
 	// scan.
-	scanList []*hostEntry
-	scanPool chan *hostEntry
+	scanList []modules.HostDBEntry
+	scanPool chan modules.HostDBEntry
 	scanWait bool
+	online   bool
 
 	blockHeight types.BlockHeight
 	lastChange  modules.ConsensusChangeID
 }
 
 // New returns a new HostDB.
-func New(cs consensusSet, persistDir string) (*HostDB, error) {
+func New(g modules.Gateway, cs consensusSet, persistDir string) (*HostDB, error) {
```

Comment:
You already know that I disagree with mocking entire dependencies for testing. You want the side effects and interfereing processes, because those effects will be there during production. If it makes the test hard to write or NDF, it's often a bug upstream, and then you are glad you found the bug. Supplying a full gateway is easy when you've got a helper function in the fashion of the contract manager testing. You ask for the root dependencies and then if you need you can overwrite some of them individually to trigger certain codepaths.

I agree that really we want a `HasInternetConnection` method as opposed to a `Peers` method. I guess we can extend the gateway to support that method.

I'm not sure exactly what you mean when you say restrictive typing, but if you are talking about redefining every dependency as a new interface which is the subset of functions that you need, I am pretty strongly against the practice. I feel like it makes testing more difficult, and it also leads you to believe that you can easily swap in some other object to satisfy the interface when the reality is that these modules are designed to depend on eachother, not on eachother's subsets.